File: app/src/main/java/com/vimers/smartblock/RegistrationActivity.java
package com.vimers.smartblock;

import android.content.Intent;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.widget.Button;
import android.widget.EditText;

import androidx.appcompat.app.AppCompatActivity;

import com.vimers.smartblock.persistence.PersistentObject;

import kotlin.Unit;

public class RegistrationActivity extends AppCompatActivity {

    private EditText mailEdit;
    private EditText passwordEdit;
    private Button completeRegistrationButton;

    private PersistentObject<AppSettings> appSettings;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_registration);

        mailEdit = findViewById(R.id.mailEdit);
        passwordEdit = findViewById(R.id.passwordEdit);
        completeRegistrationButton = findViewById(R.id.completeRegistrationBut);

        appSettings = new PersistentObject<>(this, AppSettings.class);

        TextWatcher textWatcher = new TextWatcher() {
            public void afterTextChanged(Editable s) {
                completeRegistrationButton.setEnabled(isInputDataCorrect());
            }

            private boolean isInputDataCorrect() {
                return (mailEdit.length() != 0)
                        && (passwordEdit.length() != 0);
            }

            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
            }

            public void onTextChanged(CharSequence s, int start, int before, int count) {
            }
        };

        mailEdit.addTextChangedListener(textWatcher);
        passwordEdit.addTextChangedListener(textWatcher);

        completeRegistrationButton.setOnClickListener(v -> onRegistrationCompleted());
    }

    /**
     * Saves the inputted password.
     */
    private void saveData() {
        appSettings.edit(settings -> {
            settings.setMail(mailEdit.getText().toString());
            settings.setPassword(passwordEdit.getText().toString());
            return Unit.INSTANCE;
        }).save();
    }

    private void onRegistrationCompleted() {
        saveData();
        getSharedPreferences("APP_STATE", MODE_PRIVATE).edit().putBoolean("REGISTRATION_COMPLETED", true).apply();
        startActivity(new Intent(this, SettingsActivity.class));
    }
}

File: app/src/main/java/com/vimers/smartblock/RegistrationActivity.kt
package com.vimers.smartblock

import android.content.Intent
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import com.vimers.smartblock.persistence.PersistentObject
import kotlinx.android.synthetic.main.activity_registration.*

class RegistrationActivity : AppCompatActivity() {
    private lateinit var appSettings: PersistentObject<AppSettings>

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        appSettings = PersistentObject(this, AppSettings::class.java)
        finishIfRegistrationTakenPlace()

        setContentView(R.layout.activity_registration)

        setTextWatchers()
    }

    private fun finishIfRegistrationTakenPlace() {
        if (appSettings.obj.registrationFinished)
            finish()
    }

    private fun setTextWatchers() {
        val textWatcher = object : TextWatcher {
            override fun afterTextChanged(s: Editable) {
                setFabEnabled(isInputDataCorrect)
            }

            private val isInputDataCorrect: Boolean
                get() = (emailEdit.length() != 0
                        && passwordEdit.length() != 0)

            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {}
        }
        emailEdit.addTextChangedListener(textWatcher)
        passwordEdit.addTextChangedListener(textWatcher)
    }

    private fun setFabEnabled(enabled: Boolean) {
        fab.setColorFilter(
                ContextCompat.getColor(
                        this,
                        if (enabled) R.color.colorAccentDisabled else R.color.colorAccent
                )
        )
        fab.isClickable = enabled
    }

    fun doRegistration(view: View) {
        appSettings.edit {
            mail = emailEdit.text.toString()
            password = passwordEdit.text.toString()
        }.save()
        startActivity(Intent(this, MainActivity::class.java))
    }
}